src/macaron/slsa_analyzer/package_registry/pypi_registry.py

```diff
@@ -0,0 +1,198 @@
+
+"""The module provides abstractions for the pypi package registry."""
+
+import logging
+import os
+import re
+from datetime import datetime
+from typing import Any
+
+from bs4 import BeautifulSoup
+
+from macaron.slsa_analyzer.package_registry.package_registry import PackageRegistry
+from macaron.util import send_get_http, send_get_http_raw
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+class PyPIRegistry(PackageRegistry):
+    """This class implements the npm package registry.
+
+    There is no complete and up-to-date API documentation for the npm registry and the endpoints
+    are discovered by manual inspection of links on https://www.npmjs.com.
+    """
+
+    def __init__(
+        self,
+        hostname: str | None = None,
+        attestation_endpoint: str | None = None,
+        request_timeout: int | None = None,
+        enabled: bool = True,
+    ) -> None:
+        """
+        Initialize the npm Registry instance.
+
+        Parameters
+        ----------
+        hostname : str | None
+            The hostname of the npm registry.
+        attestation_endpoint : str | None
+            The attestation REST API.
+        request_timeout : int | None
+            The timeout (in seconds) for requests made to the package registry.
+        enabled: bool
+            Shows whether making REST API calls to npm registry is enabled.
+        """
+        self.hostname = hostname or ""
+        self.attestation_endpoint = attestation_endpoint or ""
+        self.request_timeout = request_timeout or 10
+        self.enabled = enabled
+        super().__init__("pypi Registry")
+
+
+class PyPIApiClient:
+    """PyPI's official and customed API client."""
+
+    def __init__(self, package: str) -> None:
+        self.headers = {"Host": "pypi.org", "Accept": "application/json"}
+        self.base_url = "https://pypi.org"
+        self.package = package
+        self.base_metadata: dict[str, Any] = self._get_base_metadata()
+
+    def _get_base_metadata(self) -> dict:
+        """Get the metadata through PyPI official API.
+
+        Returns
+        -------
+            dict | None: PyPI official metadata.
+        """
+        endpoint = os.path.join(self.base_url, "pypi", self.package, "json")
+        response: dict = send_get_http(endpoint, headers=self.headers)
+        return response
+
+    def get_releases(self) -> dict | None:
+        """Get all releases.
+
+        Returns
+        -------
+            dict | None: Version to metadata.
+        """
+        releases: dict | None = self.base_metadata.get("releases", None)
+        return releases
+
+    def get_project_links(self) -> dict | None:
+        """Get all project links.
+
+        Returns
+        -------
+            dict | None: All link's name to link.
+        """
+        project_urls: dict | None = self.base_metadata.get("info", None).get("project_urls", None)
```

Comment:
This is not correct. Because if `self.base_metadata.get("info", None)` returns None, calling `get("project_urls", None)` on a NoneType value would raise error.
```python
>>> a = {}
>>> a.get("A").get("B")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'NoneType' object has no attribute 'get'
```

Reply:
However, if we are switch to https://github.com/oracle/macaron/blob/e214326697aee6e31f0c9e406b11344c8619581e/src/macaron/json_tools.py#L15, such issues can be prevented.

Reply:
Fixed this in f900dfef1bae0413a4474f9e8e9101db3c625af1